Tests/AttributedStringStyleTests.swift

```diff
@@ -555,6 +555,134 @@ class StringStyleTests: XCTestCase {
         XCTAssertFalse(stillHasAltSixDict)
     }
 
+    func testStyleBlockRules() {
+        let string = "0<one attr1=\"11\">1<two attr2=\"12\">2</two></one>"
+
+        var tagAttr1Value: String!
+        var tagAttr2Value: String!
+
+        let rules: [XMLStyleRule] = [
+            .styleBlock("one") { attributes in
+                guard let val1 = attributes["attr1"], let intVal1 = Int(val1) else {
+                    XCTFail("attr1 not found")
+                    return StringStyle()
+                }
+
+                tagAttr1Value = val1
+                return StringStyle(.baselineOffset(CGFloat(intVal1)))
+            },
+            .styleBlock("two") { attributes in
+                guard let val2 = attributes["attr2"], let intVal2 = Int(val2) else {
+                    XCTFail("attr2 not found")
+                    return StringStyle()
+                }
+
+                tagAttr2Value = val2
+                return StringStyle(.baselineOffset(CGFloat(intVal2)))
+            },
+        ]
+
+        let attributed = string.styled(with: .font(BONFont.fontA), .xmlRules(rules))
+        XCTAssertEqual(attributed.string, "012")
+
+        XCTAssertEqual(tagAttr1Value, "11")
+        XCTAssertEqual(tagAttr2Value, "12")
+
+        let attrs0 = attributed.attributes(at: 0, effectiveRange: nil)
+        let attrs1 = attributed.attributes(at: 1, effectiveRange: nil)
+        let attrs2 = attributed.attributes(at: 2, effectiveRange: nil)
+
+        XCTAssertTrue(attrs0.count == 1)
+        XCTAssertTrue(attrs1.count == 2)
+        XCTAssertTrue(attrs2.count == 2)
+
+        XCTAssertNil(attrs0[.baselineOffset])
+        guard let lineSpacing1 = attrs1[.baselineOffset] as? Int else { XCTFail("lineSpacing1 not found"); return }
+        guard let lineSpacing2 = attrs2[.baselineOffset] as? Int else { XCTFail("lineSpacing2 not found"); return }
+
+        XCTAssertEqual(lineSpacing1, 11)
+        XCTAssertEqual(lineSpacing2, 12)
+    }
+
+    func testEnterExitBlockRules() {
+        let string = "0<one attr1=\"11\" attr2=\"a\">1<two attr3=\"12\" attr4=\"b\">2</two></one>"
+
+        var tagAttr1Value: String!
+        var tagAttr2Value: String!
+        var tagAttr3Value: String!
+        var tagAttr4Value: String!
+
+        let rules: [XMLStyleRule] = [
+            .enterBlock(element: "one") { attributes in
+                guard let val1 = attributes["attr1"], let intVal1 = Int(val1) else {
```

Comment:
Now that I understand a little more about what's going on, I think you can use `XCTUnwrap` to simplify all lines that have this pattern. Basically anywhere you currently have a `guard`, I think an unwrap would work better (and simpler) for tests.

One caveat is that you would probably have to annotate the various blocks with `throws`, and I'm not sure how much that would complicate their use in the rest of the app, so this may be invalid advice _inside_ the blocks, but still good elsewhere in the tests.